tekton-catalog/pipeline-loops/cmd/cli/main.go

```diff
@@ -0,0 +1,232 @@
+			case "Pipeline":
+				err = validatePipeline(marshalledBytes)
+			case "PipelineLoop":
+				err = validatePipelineLoop(marshalledBytes)
+			case "PipelineRun":
+				err = validatePipelineRun(marshalledBytes)
+			default:
+				fmt.Printf("Warn: Unsupported kind: %s.\n", kind)
+			}
+			if err != nil {
+				errs = append(errs, err.Error())
+			}
+		}
+		if len(errs) > 0 {
+			fmt.Printf("Validation errors: %s\n", strings.Join(errs, "\n"))
+			os.Exit(100)
+		} else {
+			fmt.Printf("\nCongratulations, all checks passed !!\n")
+		}
+	}
+}
+
+func validatePipeline(bytes []byte) error {
+	p := v1beta1.Pipeline{}
+	err := json.Unmarshal(bytes, &p)
+	if err != nil {
+		return err
+	}
+	return validatePipelineSpec(&p.Spec, p.Name)
+}
+
+func validatePipelineSpec(p *v1beta1.PipelineSpec, name string) error {
+	errs := []string{}
+	// We do not need to validate PipelineRun because it is validated by tekton admission webhook
+	// And pipelineTask.TaskRef is also validated by tekton.
+	// Here we only need to validate those embedded spec, whose kind is pipelineLoop.
+	if p.Tasks != nil {
+		for _, task := range p.Tasks {
+			if task.TaskSpec != nil && task.TaskSpec.Kind == "PipelineLoop" {
+				err := validatePipelineLoopEmbedded(task.TaskSpec.Spec.Raw)
+				if err != nil {
+					errs = append(errs, err.Error())
+				}
+			}
+		}
+	}
+	if len(errs) > 0 {
+		e := strings.Join(errs, "\n")
+		return fmt.Errorf("Validation errors found in pipeline %s\n %s", name, e)
+	}
+	return nil
+}
+
+func validateRun(bytes []byte) error {
+	r := v1alpha1.Run{}
+	err := json.Unmarshal(bytes, &r)
+	if err != nil {
+		return err
+	}
+	// We do not need to validate Run because it is validated by tekton admission webhook
+	// And r.Spec.Ref is also validated by tekton.
+	// Here we only need to validate the embedded spec. i.e. r.Spec.Spec
+	if r.Spec.Spec != nil && r.Spec.Spec.Kind == "PipelineLoop" {
+		if err := validatePipelineLoopEmbedded(r.Spec.Spec.Spec.Raw); err != nil {
+			return fmt.Errorf("Found validation errors in Run: %s \n %s", r.Name, err.Error())
+		}
+	}
+	return nil
+}
+
+func validatePipelineRun(bytes []byte) error {
+	pr := v1beta1.PipelineRun{}
+	err := json.Unmarshal(bytes, &pr)
+	if err != nil {
+		return err
+	}
+	return validatePipelineSpec(pr.Spec.PipelineSpec, pr.Name)
+}
+
+func validatePipelineLoopEmbedded(bytes []byte) error {
+	var embeddedSpec map[string]interface{}
+	if err := json.Unmarshal(bytes, &embeddedSpec); err != nil {
+		return err
+	}
+	r1 := map[string]interface{}{
+		"kind":       "PipelineLoop",
+		"apiVersion": "custom.tekton.dev/v1alpha1",
+		"metadata":   metav1.ObjectMeta{Name: "embedded"},
+		"spec":       embeddedSpec,
+	}
+
+	marshalBytes, err := json.Marshal(r1)
+	if err != nil {
+		return err
+	}
+	return validatePipelineLoop(marshalBytes)
+}
+
+func validatePipelineLoop(bytes []byte) error {
+	pipelineLoop := pipelineloopv1alpha1.PipelineLoop{}
+	if err := json.Unmarshal(bytes, &pipelineLoop); err != nil {
```

Comment:
I am a little confuse here. 
If we want to use this tool for pipelineloop validate, then the iterateNumeric or iterateParam must be exist, and also the spec.pipelinespec.tasks.
If we want to use this tool for common Tekton custom task check, then the fuction name of pipelineloop related words need to be replaced.